src/java_bytecode/java_static_initializers.cpp

```diff
@@ -134,82 +260,239 @@ static void create_clinit_wrapper_symbols(
     "clinit wrapper");
 }
 
-/// Produces the static initialiser wrapper body for the given function.
+/// Produces the static initialiser wrapper body for the given function. This
+/// static initializer implements (a simplification of) the algorithm defined
+/// in Section 5.5 of the JVM Specs. This function, or wrapper, checks whether
+/// static init has already taken place, calls the actual `<clinit>` method if
+/// not, and possibly recursively initializes super-classes and interfaces.
+///
+/// ```
+///   bool was_ready;
+///   if(java::C::__CPROVER_clinit_local_state == RETURN)
+///   {
+///     return;
+///   }
+///   java::C::__CPROVER_clinit_local_state = RETURN;
+///
+///   ATOMIC_BEGIN
+///   Assume(java::C::__CPROVER_clinit_state != IN_PROGRESS)
```

Comment:
?? Will that work? Presumably we want the other thread to busy-wait for IN_PROGRESS to transition to DONE in this case. If this is indeed magic that symex understands like a wait then write that in the comments.

Reply:
Yes, that will work. It's exactly how POSIX mutexes are implemented in `pthread_mutex_lock`, see here:

https://github.com/diffblue/cbmc/blob/develop/src/ansi-c/library/pthread_lib.c#L132

Thanks, we will indeed document it.